app/scripts/metamask-controller.js

```diff
@@ -3327,6 +3326,17 @@ export default class MetamaskController extends EventEmitter {
    */
   async connectHardware(deviceName, page, hdPath) {
     const keyring = await this.getKeyringForDevice(deviceName, hdPath);
+
+    if (deviceName === HardwareDeviceNames.qr) {
+      keyring
+        .getMemStore()
+        .subscribe(
```

Comment:
Hmm, I can't find anyplace where this subscription is removed. What happens when the wallet is locked and unlocked again?

Before this PR, there was a single QR Keying instance that was never destroyed (which for keyrings is quite atypical). Now that we're treating it more like a normal keyring, we'll have to handle destroying it as well.

Reply:
This seems... tricky. The same problem applies for when the keyring gets created too - this subcription only gets setup for once instance, but when it's destroyed and recreated again after lock/unlock it'll stop updating.

A lot of the QR keyring related methods in core will implicitly create a QR keyring right now, as will unlock.

And for the destroy case, we have lock and forget keyring.

Reply:
Good catch! 
Perhaps the KeyringController should make this subscription when adding the QRKeyring (also in case of unlock).

For the missing unsubscribe: When the QR gets forgotten it is not removed - we call `keyring.forgetDevice`, which reinitializes the keyring state, so the subscription keeps working as the QR instance remains in memory.

When locking though, the QRKeyring gets unreferenced from the EthKeyringController's keyrings array, which should cause the subscription to not be called again (since events are fired from the ObservableStore of the QRKeyring instance, which becomes unreferenced)

Reply:
>Perhaps the KeyringController should make this subscription when adding the QRKeyring (also in case of unlock), and expose its update events through the controller messenger

This sounds like a good solution to me!

Reply:
@Gudahtt  Added in https://github.com/MetaMask/core/pull/1702